### PR TITLE
Add "Add Contact" functionality to Saros/I

### PR DESCRIPTION
**Pull request by tobous**
#### [INTERNAL][I] Re-draw contacts view when contacts are added or removed

Re-draws the contact view when a contact is added or removed.

Adjusts the re-drawing logic to do nothing if the project the component
is displayed in is disposed, meaning it was closed (or is currently in
the process of being closed). This was done to avoid unnecessary update
option and avoid possible issues with interacting with the UI while it
is being torn down.

#### [FEATURE][I] Add "Add Contact" dialog

Adds the button and corresponding dialog to add a contact to the
currently connected XMPP account. The button is only enabled when there
is a valid connection to an XMPP account.

Adjusts the text of the dialog to add an XMPP account to the account
store to stay consistent. The text was adjusted to form a request
instead of a statement. This was done to avoid confusion as the dialog
title is not always displayed (dependent on operating system/desktop
environment).

### CODE REVIEW

**Comment by tobous** [comment]
This looks like a lot more than it actually is. A lot of it are basic refactorings and the usage of basic UI logic.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  6
         

Complexity decreasing per file
==============================
+ intellij/src/saros/intellij/ui/views/SarosToolbar.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4659323&bid=15602194)

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  /**
+   * Shows the dialogs requesting the JID and (optional) nickname for the account to add as a
+   * contact.
+   *
+   * <p>Calls {@link XMPPContactsService#addContact(JID, String, BiPredicate)} if the dialogs were
+   * not canceled.
+   */
+  private void showAddContactDialog() {
```
Comment (tobous) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Will split the retrieval of the jid and nickname off into separate methods.

File: intellij/src/saros/intellij/ui/tree/SessionAndContactsTreeView.java [hunk 2]
```diff
@@ -113,6 +126,7 @@ public SessionAndContactsTreeView(@NotNull Project project) {
   @Override
   public void dispose() {
     connectionHandler.removeConnectionStateListener(connectionStateListener);
+    xmppContactsService.removeListener(contactsUpdateListener);
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/tree/SessionAndContactsTreeView.java]: Isnt this widget missing some kind of isDisposed() handling ? The events could be triggered while a dispose is in progress,

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    if (nickname == null) {
+      log.debug("Adding contact canceled by user during user nickname entry.");
+
+      return;
+    }
+
+    BiPredicate<String, String> dialog =
+        (title, message) -> {
+          log.fatal("showing dialog " + title + " - " + message);
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: ?

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+          Integer integer;
+
+          try {
+            integer = SafeDialogUtils.showYesNoDialog(project, message, title);
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: I do not know if it is a good idea to return null as an integer here ? And when is null returned ?

Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: It can't return null (unless the dialog aborts due to an internal error?). The usage of `Integer` stems from the dialog being executed synchronously in a runnable. [Here is the code.](https://github.com/saros-project/saros/blob/9b13dea8df827a2aea8d7b2b52b1c763d6f6040b/intellij/src/saros/intellij/ui/util/SafeDialogUtils.java#L203) To fetch the result, an `AtomicReference<>` is used, which requires an object type. But keeping it as an Integer for the return value is unnecessary.

That is an oversight on my part. It should unpack the int (or, even better, do the check against `Messages#YES` and `Messages#No` and return a `boolean`). Will adjust in a separate patch.

File: intellij/src/saros/intellij/ui/views/buttons/AbstractToolbarButton.java [hunk 5]
```diff
@@ -1,33 +1,74 @@
...
-    UIUtil.invokeAndWaitIfNeeded((Runnable) () -> setEnabled(enabled));
+  /**
+   * Calls {@link AbstractButton#setEnabled(boolean)} on the UI thread.
+   *
+   * @param enabled <code>true</code> to enable the button, <code>false</code> to disable the button
+   * @see UIUtil#invokeAndWaitIfNeeded(Runnable)
+   */
+  @Override
+  public void setEnabled(boolean enabled) {
+    UIUtil.invokeAndWaitIfNeeded((Runnable) () -> super.setEnabled(enabled));
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AbstractToolbarButton.java]: Using such syncExec calls gave us a lot of dead locks in Eclipse/E. I do not even know if it is good to coordinate the access to the widget here however I suggest to use ```invokeLaterIfNeeded``` instead.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+          @Override
+          public boolean checkInput(String inputString) {
+            return true;
+          }
+
+          @Override
+          public boolean canClose(String inputString) {
+            JID jid = new JID(inputString);
+
+            return jid.isValid() && !jid.getName().isEmpty();
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Reading the IntelliJ Javadoc is not really helpful here. What is displayed when this method returns false ? Is there any hint ?

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+          }
+        };
+
+    String userID;
+
+    try {
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: This is not very user friendly. First you ask for the JID and afterwards another dialog is opened that asks for an optional nickname ? Any reasons why this is different from Eclipse where you can insert the JID and nickname in the same wizard/dialog page ?

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+          SafeDialogUtils.showInputDialog(
+              project,
+              Messages.AddContactButton_contact_jid_dialog_message,
+              Messages.AddContactButton_contact_jid_dialog_initial_input,
+              Messages.AddContactButton_contact_jid_dialog_title,
+              inputValidator,
+              new TextRange(0, 0));
+
+    } catch (IllegalAWTContextException e) {
+      log.error("Account creation failed.", e);
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: The account creation has not even started here. Can this exception even happen on normal user interaction. Reading the Javadoc for the exception this could only happen if we somehow manage to call this code inside a write access which currently seems impossible for me.

TLDR: I guess the TODO can be removed.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    BiPredicate<String, String> dialog =
+        (title, message) -> {
+          log.fatal("showing dialog " + title + " - " + message);
+
+          Integer integer;
+
+          try {
+            integer = SafeDialogUtils.showYesNoDialog(project, message, title);
+          } catch (IllegalAWTContextException e) {
+            log.error("Tried to run dialog in an illegal context", e);
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Cannot happen here as it would already happen on line 123.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+          }
+
+          return integer != null && integer.equals(com.intellij.openapi.ui.Messages.YES);
+        };
+
+    try {
+      xmppContactsService.addContact(jid, nickname, dialog);
+    } catch (OperationCanceledException e) {
+      // TODO
+      log.error("Adding contact failed", e);
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: I do not think that this is an error here. The user has canceled on purpose or did I miss some details what is contained in the exception ?

File: intellij/src/saros/intellij/ui/views/buttons/ConsistencyButton.java [hunk 11]
```diff
@@ -149,11 +151,11 @@ private void setInconsistent(boolean isInconsistent) {
     sessionInconsistencyState.isInconsistent = isInconsistent;
 
     if (isInconsistent) {
-      setEnabledFromUIThread(true);
+      setEnabled(true);
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/ConsistencyButton.java]: I do not know the IntelliJ/AWT/Swing guidelines but it is pretty weird that this call is explicitly synchronized with the UI Thread but setButtonIcon and setToolTipText is not (I do not know the internal implementation but your setEnabled override just calls the original setEnabled method. However you did not do that for setButtonIcon and setToolTipText).

File: intellij/src/saros/intellij/ui/views/buttons/SimpleButton.java [hunk 12]
```diff
@@ -1,17 +1,30 @@
 package saros.intellij.ui.views.buttons;
 
+import com.intellij.openapi.project.Project;
 import javax.swing.ImageIcon;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 import saros.intellij.ui.actions.AbstractSarosAction;
 
 /** Simple button used to create actions that just call {@link AbstractSarosAction#execute()}. */
 public class SimpleButton extends AbstractToolbarButton {
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/SimpleButton.java]: Good candidate for renaming. Isn't ActionToolbarButton a better name ?

File: intellij/src/saros/intellij/ui/views/buttons/AbstractSessionToolbarButton.java [hunk 13]
```diff
@@ -21,13 +19,12 @@
...
  *
- * <p><b>NOTE:</b>This component and any component added here must be correctly torn down when the
- * project the components belong to is closed. See {@link SarosMainPanelView}. This class offers the
- * method {@link #disposeComponents()} for such a purpose.
+ * <p>Implementing classes <b>must</b> call <code>super.dispose()</code> when overwriting the {@link
+ * #dispose()} method.
+ *
+ * @see AbstractToolbarButton
  */
 abstract class AbstractSessionToolbarButton extends AbstractToolbarButton implements Disposable {
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AbstractSessionToolbarButton.java]: The Interface is already inherited by inheriting from AbstractToolbarButton

File: intellij/src/saros/intellij/ui/views/buttons/AbstractToolbarButton.java (hunk 5 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AbstractToolbarButton.java]: To be honest, I just made this change to be consistent. Both this and the default `setEnabled()` were used and I thought if someone bothered to add it there surely would be a need/purpose for it.

From reading the IntelliJ documentation on GUI interactions, I couldn't find any specification on whether to enable components synchronously. (And IntelliJ doesn't actively seem to mind as calls to the default `setEnabled()` work just fine.) But I am not well versed in the are of GUI programming. So if there are any other reasons to do this synchronously in a dedicated thread, please tell me.

Otherwise, I would simply adjust the corresponding commit by removing the `setEnabled()` implementation and instead using the default one.

File: intellij/src/saros/intellij/ui/views/buttons/AbstractToolbarButton.java [hunk 14]
```diff
@@ -1,33 +1,74 @@
...
   }
 
+  /**
+   * Method to dispose components of the button.
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/views/buttons/AbstractToolbarButton.java]: I suggest to use inheriteddoc and amend some specific content or just drop the whole declaration.

You could also give an example here. E.g unregister listeners etc.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java (hunk 7 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: This is how the current user dialog is mostly handled (account addition works this way as well; the incoming session and project negotiation are the only components that currently use the wizard).

Theoretically, this could be implemented using the Saros/I wizard, but the current implementation (located [here](https://github.com/saros-project/saros/tree/master/intellij/src/saros/intellij/ui/wizards) is not very good from what I can tell. It
- does not use the IntelliJ swing wrappers consistently,
- is somewhat difficult to work with when implementing new pages (as basic elements of the wizard, like control over the buttons, are not accessible to the pages (as far as I know), making it hard to provide a well-rounded user interaction), and
- lacks basic UI features like "automatic window scaling to content size" and "content size scaling to window size" (i.e. it plain looks bad and most of the time it does not have the right size for its displayed content, leading to cut-off text or a *lot* of blank space).

So for the quick and dirty implementation of the add contact dialog, I chose to use the default dialogs.

As I already mentioned multiple times, the UI is still one of the most outdated/broken/lackluster aspects of the Saros/I implementation and would need a lot more attention. But, as it *kind of* works currently, it seems like a lower priority than providing basic Saros features to me (and I am not used to/don't like implementing GUI logic 😉).

And even with the lacking quality of the existing components, when working on the GUI, I would still put implementing new features (like the settings) before improving the existing ones (unless they are related). Either way, I will create issues for the missing GUI features in Saros/I (and will probably also include the stuff that needs improvements). Then we can maybe get some outside collaborators (or students) to work on that as it should mostly be independent from the main Saros logic, meaning not that much base knowledge of Saros is required (so it is easier to get going) and the influence on other components is limited.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java (hunk 8 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Well, the user pressed the button and nothing happened. So it would be nice to have some kind of feedback (especially as it will enable the user to make a more accurate bug report than just "nothing happened when I pressed the button"). But yes, the log message is misleading.

Regarding whether this can actually happen: I don't think so, but I don't think it hurts to cover all bases in this case.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java (hunk 10 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: > The user has canceled on purpose or did I miss some details what is contained in the exception ?

Nope. I already adjusted this on my local version. I just wrote that before actually reading what the exception was coming from. Will just debug log and exit quietly.

File: intellij/src/saros/intellij/ui/views/buttons/ConsistencyButton.java (hunk 11 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/ConsistencyButton.java]: As mentioned in the comment on the actual implementation, I just assumed that the pre-existing method to synchronize the enable access was put there on purpose/for a reason. It can probably be removed.

File: intellij/src/saros/intellij/ui/views/buttons/SimpleButton.java (hunk 12 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/SimpleButton.java]: Probably. Also, it is a good canidate for removal. It was basically only used to disaply buttons without an implementation, which are currently not shown anyways. So just throwing it away and replacing the usage with a TODO "add this button" would be for the best in my opinion.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java (hunk 9 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Yeah, but I need to catch it anyway. And I will most likely handle the user reporting in a separate method, so it shouldn't add much code to handle this case the same as the first call.

**Comment by tobous** [comment]
As the refactoring part of the PR was starting to grow larger than the actual changes the PR was meant for, I decided to move them to a dedicated PR #789  that is now used as the base for this PR. I addressed/fixed the issues regarding the refactoring mentioned here and marked the corresponding comments as resolved beforehand.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4659323&bid=15602194)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4659323&bid=15602194)

File: intellij/src/saros/intellij/ui/tree/SessionAndContactsTreeView.java [hunk 15]
```diff
@@ -105,6 +111,21 @@ public SessionAndContactsTreeView(@NotNull Project project) {
 
     connectionHandler.addConnectionStateListener(connectionStateListener);
 
+    contactsUpdateListener =
+        (contact, updateType) -> {
+          if (project.isDisposed()) {
```
Comment (srossbach) [inline intellij/src/saros/intellij/ui/tree/SessionAndContactsTreeView.java]: I do not know the threading requirements but such logic would not work in SWT. 

This is not called by the UI thread (which basically does a dispose) and so even if this returns false the dispose may happen just after the check. I guess Swing is more graceful in handling this but you would be surprised what can happen in SWT when using this pattern.

Comment (tobous) [reply intellij/src/saros/intellij/ui/tree/SessionAndContactsTreeView.java]: Hmmmm. The only alternative I see would be to move the check into the `UIUtil.invokeLaterIfNeeded(...)` block in `renderConnected()` as I would assume that the project disposal is also run on the EDT. That would ensure that the action is aborted if the disposal is run first.

Also, this shouldn't really hurt either way. The logic called in the listener just updates the Saros view, which should not be dependent on any project specific content but rather should just deal with the Saros state. And, after the action took place, the objects should still be garbage-collectible either way.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4659323&bid=15602194)

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java (hunk 6 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Added javadoc stating what the method does. Also added an error dialog that is shown if the user presses "OK" while having entered an invalid JID. This gives feedback on why the dialog did not close.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4659323&bid=15602194)

**Comment by tobous** [comment]
Addressed remaining review comments and marked PR as ready for review (i.e. no longer a WIP).

**Comment by tobous** [comment]
Rebased onto current master without any interaction.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4659323&bid=15602194)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4659323&bid=15602194)

**Comment by tobous** [comment]
Build failures are unrelated to this PR. See #795.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=4659323&bid=15602194)

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+public class AddContactButton extends AbstractToolbarButton {
+  private static final Logger log = Logger.getLogger(AddContactButton.class);
+
+  @Inject private XMPPContactsService xmppContactsService;
+  @Inject private ConnectionHandler connectionHandler;
+
+  private final IConnectionStateListener connectionStateListener =
+      (state, error) -> {
+        if (state == ConnectionState.CONNECTED) {
+          setEnabled(true);
```
Comment (stefaus) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Nit: alternative `setEnabled(state == ConnectionState.CONNECTED)`

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+   * @param nickname the nickname to use or an empty string if no nickname was specified
+   * @see XMPPContactsService#addContact(JID, String, BiPredicate)
+   */
+  private void addContact(@NotNull JID jid, @NotNull String nickname) {
+    BiPredicate<String, String> dialog =
+        (title, message) -> {
+          boolean choseYes;
+
+          try {
+            choseYes = SafeDialogUtils.showYesNoDialog(project, message, title);
```
Comment (stefaus) [inline intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Nit: `return SafeDialogUtils.showYesNoDialog(project, message, title);`

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java (hunk 16 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Done.

File: intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java (hunk 17 above)
Comment (tobous) [reply intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java]: Done.

**Comment by tobous** [comment]
Addressed review comments and rebased onto current master.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- intellij/src/saros/intellij/ui/views/buttons/AddContactButton.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=4659323&bid=15602194)
